src/metabase/sync/fetch_metadata.clj

```diff
@@ -28,25 +30,56 @@
-      (set (fields-metadata database :table-names [(:name table)] :schema-names [(:schema table)]))
-      (:fields (driver/describe-table (driver.u/database->driver database) database table)))))
+    (let [driver (driver.u/database->driver database)
+          result (if (driver/database-supports? driver :describe-fields database)
+                   (set (driver/describe-fields driver
+                                                database
+                                                :table-names [(:name table)]
+                                                :schema-names [(:schema table)]))
+                   (:fields (driver/describe-table driver database table)))]
+      (cond-> result
+        (driver/database-supports? driver :nested-field-columns database)
+        ;; TODO: decouple nested field columns sync from field sync. This will allow
+        ;; describe-fields to be used for field sync for databases with nested field columns
+        ;; Also this should be a driver method, not a sql-jdbc.sync method
+        (set/union (sql-jdbc.sync/describe-nested-field-columns driver database table))))))
+
+(defn- describe-fields-using-describe-table
+  "Replaces [[metabase.driver/describe-fields]] for drivers that haven't implemented it. Uses [[driver/describe-table]]
+  instead. Also includes nested field column metadata."
+  [_driver database & {:keys [schema-names table-names]}]
+  (let [tables (sync-util/db->reducible-sync-tables database :schema-names schema-names :table-names table-names)]
+    (eduction
+     (mapcat (fn [table]
+               (for [x (table-fields-metadata database (t2.realize/realize table))]
+                 (assoc x :table-schema (:schema table) :table-name (:name table)))))
+     tables)))
 
-(defn backwards-compatible-describe-fks
+(mu/defn fields-metadata
+  "Effectively a wrapper for [[metabase.driver/describe-fields]] that also validates the output against the schema.
+  If the driver doesn't support [[metabase.driver/describe-fields]] it uses [[driver/describe-table]] instead.
+  This will be deprecated in "
+  [database :- i/DatabaseInstance & {:as args}]
+  (log-if-error "fields-metadata"
+    (let [driver             (driver.u/database->driver database)
+          describe-fields-fn (if (driver/database-supports? driver :describe-fields database)
+                               driver/describe-fields
+                               ;; In a future version we may remove [[driver/describe-table]]
+                               ;; and we'll just use [[driver/describe-fields]] here
+                               describe-fields-using-describe-table)]
+      (cond->> (describe-fields-fn driver database args)
+        ;; This is a workaround for the fact that [[mu/defn]] can't check reducible collections yet
+        (mu.fn/instrument-ns? *ns*)
+        (eduction (map #(mu.fn/validate-output {} i/FieldMetadataEntry %)))))))
+
+(defn- describe-fks-using-describe-table-fks
```

Comment:
I renamed this from `backwards-compatible-describe-fks` to make it consistent with `describe-fields-using-describe-table`.